### PR TITLE
Feature/profile image uploads

### PR DESCRIPTION
**Pull request by crobbo**
/claim #674

[Screencast from 2024-04-27 10-13-24.webm](https://github.com/maybe-finance/maybe/assets/47796704/7e2fc296-ddf8-48af-a2f3-437b410e5fab)

### CODE REVIEW

File: config/storage.yml
Comment (zachgoll) [inline config/storage.yml:27]: ERB syntax error here

File: config/storage.yml
Comment (zachgoll) [inline config/storage.yml:20]: We should probably set a more descriptive keyfile name such as `gcp-storage-keyfile.json` and then add that to the .gitignore here

File: Gemfile
Comment (zachgoll) [inline Gemfile:41]: Is there a specific reason for using a different versions from the Rails guides for AWS and Azure?

```
gem 'aws-sdk-s3', require: false
gem 'azure-storage-blob', "~> 2.0", require: false
```

**Review by zachgoll — COMMENTED** [review]
The upload and preview flow looks good.  Left some more specific comments on improvements we could make.

In addition to those comments, can we also add a section in `.env.example` that explains how to configure this and what env vars are available for use?  (similar to what we have for other self-host configs)

File: Gemfile
Comment (crobbo) [reply Gemfile:41]: Nope. I missed that in the guide.

File: app/views/settings/profiles/show.html.erb
Comment (crobbo) [reply app/views/settings/profiles/show.html.erb:24]: Amended this so it deletes the image. 

It always irritates me when an app lets you upload an image but not remove it, so you have to upload a new image to replace it.

File: app/models/user.rb
Comment (crobbo) [reply app/models/user.rb:12]: That seems a sensible thing to include.

**Comment by crobbo** [comment]
@zachgoll thanks for the review, it was all valid feedback. I've made those amends and requested a re-review 👍🏻

File: .env.example
Comment (zachgoll) [inline .env.example:88]: Perfect, looks good 👍

File: app/controllers/settings/profiles_controller.rb
Comment (zachgoll) [inline app/controllers/settings/profiles_controller.rb:10]: Since this form is tied to the update for a user's first and last name, the profile picture will be cleared out on every save to the form.

Repro:

1. Add picture
2. Hit save
3. Hit save again

File: app/models/user.rb
Comment (zachgoll) [inline app/models/user.rb:22]: 2 notes here:

1. Can we get this added to the sidebar and within settings menu?

CleanShot 2024-04-29 at 09 42 45


3. We could probably add `preprocessed: true` here since this shows in sidebar and we're confident it will always need to load

**Review by zachgoll — COMMENTED** [review]
Thanks for the changes here, looks like there are just a few remaining tasks here before merging:

1. Make sure lint / test all passes (i'd run these locally with `bin/rails test:all` and `bin/erblint ./app/**/*.erb` since first-time contributors have to wait on approval for GH workflows to run)
2. Fix bug where profile deletes on second save of profile
3. Add profile image to sidebar

File: config/environments/production.rb
Comment (zachgoll) [inline config/environments/production.rb:34]: I _think_ this will fail entirely if `ACTIVE_STORAGE_SERVICE` is not defined.  Probably better to use `ENV.fetch`

File: app/controllers/settings/profiles_controller.rb
Comment (crobbo) [reply app/controllers/settings/profiles_controller.rb:10]: Oh good spot.

File: app/models/user.rb
Comment (crobbo) [reply app/models/user.rb:22]: Done this but not sure what where I would add `preprocessed: true`.

How about adding `.processed` to the model method?

```ruby
  def profile_image_thumbnail
    profile_image.variant(resize_to_limit: [150, 150]).processed
  end
  ```

Comment (zachgoll) [reply app/models/user.rb:22]: I was just referring to the Rails guide example here:

https://guides.rubyonrails.org/active_storage_overview.html#:~:text=If%20you%20know%20in%20advance%20that%20your%20variants%20will%20be%20accessed%2C%20you%20can%20specify%20that%20Rails%20should%20generate%20them%20ahead%20of%20time%3A

File: app/views/shared/_user_profile_image.html.erb
Comment (zachgoll) [inline app/views/shared/_user_profile_image.html.erb:1]: I think we'll need a `h-full w-full` here:

CleanShot 2024-04-29 at 17 04 29

Comment (crobbo) [reply app/views/shared/_user_profile_image.html.erb]: Good spot I was testing with a perfectly square image.  Added `object-cover` too to prevent the image looking warped. Also form validations for image format which were missed!

File: app/models/user.rb
Comment (crobbo) [reply app/models/user.rb]: This is a better way to do it. Nice one 👍🏻

**Comment by zachgoll** [comment]
@crobbo I think all the code looks good now.  Not entirely sure what's going on with the CI checks—you _may_ need to revert back to the lockfile on `main` and call `bundle install` again since we've merged some dependabot changes recently

**Comment by crobbo** [comment]
@zachgoll possibly a bad merge somewhere. Think that last commit should fix the CI issue.

**Comment by zachgoll** [comment]
@crobbo checks are working now.  Just need to resolve lint errors and we'll be set.

**Comment by crobbo** [comment]
@zachgoll oh there's a rubocop linter too, done 👍🏻

**Comment by zachgoll** [comment]
@crobbo I merged another PR not thinking about the merge conflicts that would be introduced here.  Sorry about that, would you mind resolving those conflicts real quick and then I'll get this in?

**Comment by crobbo** [comment]
@zachgoll no problem, have done this 👍🏻

**Review by zachgoll — APPROVED** [review]
Looks good, thanks for the work on this!  Will get it merged and the bounty should be paid out automatically.

**Pull request merged**
